cores/esp32/HWCDC.cpp

```diff
@@ -119,45 +117,138 @@ static void hw_cdc_isr_handler(void *arg) {
                 break;
             }
         }
-        //send event?
-        //ets_printf("RX:%u/%u\n", i, rx_fifo_len);
         event.rx.len = i;
         arduino_hw_cdc_event_post(ARDUINO_HW_CDC_EVENTS, ARDUINO_HW_CDC_RX_EVENT, &event, sizeof(arduino_hw_cdc_event_data_t), &xTaskWoken);
+        connected = true;
     }
 
     if (usbjtag_intr_status & USB_SERIAL_JTAG_INTR_BUS_RESET) {
         usb_serial_jtag_ll_clr_intsts_mask(USB_SERIAL_JTAG_INTR_BUS_RESET);
-        initial_empty = false;
-        usb_serial_jtag_ll_ena_intr_mask(USB_SERIAL_JTAG_INTR_SERIAL_IN_EMPTY);
-        //ets_printf("BUS_RESET\n");
         arduino_hw_cdc_event_post(ARDUINO_HW_CDC_EVENTS, ARDUINO_HW_CDC_BUS_RESET_EVENT, &event, sizeof(arduino_hw_cdc_event_data_t), &xTaskWoken);
+        connected = false;
+    }
+
+    if (usbjtag_intr_status & USB_SERIAL_JTAG_INTR_SOF) {
+        usb_serial_jtag_ll_clr_intsts_mask(USB_SERIAL_JTAG_INTR_SOF);
+        lastSOF_ms = millis();
     }
 
     if (xTaskWoken == pdTRUE) {
         portYIELD_FROM_ISR();
     }
 }
 
+inline bool HWCDC::isPlugged(void)
+{
+    return (lastSOF_ms + SOF_TIMEOUT) >= millis();
+}
+
+bool HWCDC::isCDC_Connected() 
+{
+    static bool running = false;
+
+    // USB may be unplugged
+     if (!isPlugged()) {
+        connected = false;
+        running = false;
+        SOF_TIMEOUT = 5;       // SOF timeout when unplugged
+        return false;
+    } else {
+        SOF_TIMEOUT = 50;      // SOF timeout when plugged
+    }
+
+    if (connected) {
+        running = false;
+        return true;
+    }
+
+    if (running == false && !connected) { // enables it only once!
+        usb_serial_jtag_ll_ena_intr_mask(USB_SERIAL_JTAG_INTR_SERIAL_IN_EMPTY);
+    }
+    // this will feed CDC TX FIFO to trigger IN_EMPTY 
+    //uint8_t c = '\0';
+    //usb_serial_jtag_ll_write_txfifo(&c, sizeof(c));
```

Comment:
is there a reason to keep the code commented here?

Reply:
I was testing if that code was really necessary. It looks like I can just remove it.